pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1205 @@ func (o *ClusterUninstaller) validate() error {
-		return true
+	parsed, err := arn.Parse(arnString)
+	if err != nil {
+		return err
 	}
 
-	for _, tg := range filteredTGs {
-		logger.Debugf("Deleting target groups: %v", *tg.TargetGroupName)
-		_, err := elbv2Client.DeleteTargetGroup(&elbv2.DeleteTargetGroupInput{
-			TargetGroupArn: tg.TargetGroupArn,
-		})
-		if err != nil {
-			logger.Debugf("Error deleting target groups %v: %v", *tg.TargetGroupName, err)
-		} else {
-			logger.WithField("name", *tg.TargetGroupName).Info("Deleted target group")
-		}
+	switch parsed.Service {
+	case "ec2":
+		return deleteEC2(session, parsed, logger)
+	case "elasticloadbalancing":
+		return deleteElasticLoadBalancing(session, parsed, logger)
+	case "iam":
+		return deleteIAM(session, parsed, logger)
+	case "route53":
+		return deleteRoute53(session, parsed, logger)
+	case "s3":
+		return deleteS3(session, parsed, logger)
+	default:
+		return errors.Errorf("unrecognized ARN service %s (%s)", parsed.Service, arnString)
 	}
-	return false
 }
 
-// rtHasMainAssociation will check whether a given route table has an association marked 'Main'
-func rtHasMainAssociation(rt *ec2.RouteTable) bool {
-	for _, association := range rt.Associations {
-		if *association.Main == true {
-			return true
-		}
+func deleteEC2(session *session.Session, arn arn.ARN, logger logrus.FieldLogger) error {
+	client := ec2.New(session)
+
+	segments := strings.SplitN(arn.Resource, "/", 2)
+	if len(segments) != 2 {
+		return errors.Errorf("cannot parse resource %q into {type}/{id}", arn.Resource)
+	}
+	resourceType := segments[0]
+	id := segments[1]
+	logger = logger.WithField("id", id)
+
+	switch resourceType {
+	case "elastic-ip":
+		return deleteEC2ElasticIP(client, id, logger)
+	case "instance":
+		return deleteEC2Instance(client, iam.New(session), id, logger)
+	case "internet-gateway":
+		return deleteEC2InternetGateway(client, id, logger)
+	case "natgateway":
+		return deleteEC2NATGateway(client, id, logger)
+	case "route-table":
+		return deleteEC2RouteTable(client, id, logger)
+	case "security-group":
+		return deleteEC2SecurityGroup(client, id, logger)
+	case "subnet":
+		return deleteEC2Subnet(client, id, logger)
+	case "volume":
+		return deleteEC2Volume(client, id, logger)
+	case "vpc":
+		return deleteEC2VPC(client, elb.New(session), elbv2.New(session), id, logger)
+	default:
+		return errors.Errorf("unrecognized EC2 resource type %s", resourceType)
 	}
-	return false
 }
 
-// deleteVPCEndpoints will find all VPC endpoints associated with the passed in VPC and attempt to delete them
-func deleteVPCEndpoints(vpc *ec2.Vpc, ec2Client *ec2.EC2, logger logrus.FieldLogger) error {
-	describeEndpointsInput := ec2.DescribeVpcEndpointsInput{}
-	describeEndpointsInput.Filters = []*ec2.Filter{
-		{
-			Name:   aws.String("vpc-id"),
-			Values: []*string{vpc.VpcId},
-		},
-	}
-
-	results, err := ec2Client.DescribeVpcEndpoints(&describeEndpointsInput)
+func deleteEC2ElasticIP(client *ec2.EC2, id string, logger logrus.FieldLogger) error {
```

Comment:
This is different as we are not deleting network interface before disassociating https://github.com/openshift/installer/blob/d1272424396f1d42faf239dc1b976db905d0b338/pkg/destroy/aws/aws.go#L615

Reply:
> This is different as we are not deleting network interface before disassociating...

That didn't seem necessary to me, because these network interfaces are associated with load balancers, instances, and other things we have code to directly delete.  And the network interfaces don't block deleting those resources.  The fact that this code successfully removes EIPs without blocking on them being returned by the tag search should be sufficient to show this, but I can add a note to the commit message discussing if you like.  Thoughts?

Reply:
> Thoughts?

i think we should put that in commit message.